src/snowflake/snowpark/lineage.py

```diff
@@ -0,0 +1,454 @@
+
+        object_domain = self.user_to_system_domain_map.get(
+            object_domain.lower(), object_domain
+        )
+
+        parent_param = ""
+        query_object_name = object_name
+        if object_version:
+            parent_param = f', {_ObjectField.PARENT_NAME}:"{object_name}"'
+            query_object_name = object_version
+
+        query_template = '{{{nodeKey}({domainKey}: {domain}, {nameKey}:"{name}"{parent_param}) {{{edges}}}}}'
+        query = query_template.format(
+            nodeKey=_DGQLFields.NODE,
+            domainKey=_ObjectField.DOMAIN,
+            nameKey=_ObjectField.NAME,
+            domain=object_domain.upper(),
+            name=query_object_name,
+            parent_param=parent_param,
+            edges="".join(parts),
+        )
+
+        return f"select SYSTEM$DGQL('{query}')"
+
+    def _get_lineage(
+        self,
+        object_name: str,
+        object_domain: str,
+        directions: list,
+        object_version: Optional[str] = None,
+        current_depth=1,
+    ) -> List[Tuple[VariantType, VariantType, StringType, int]]:
+        """
+        Constructs and executes a query to trace the lineage of a given entity at a single depth level.
+        """
+        query_string = self._build_graphql_query(
+            object_name, object_domain, directions, object_version
+        )
+        response = self._session.sql(query_string)
+        json_response = json.loads(response.collect()[0][0])
+
+        rows = []
+        for direction in directions:
+            edges = (
+                json_response.get(_DGQLFields.DATA, {})
+                .get(_DGQLFields.NODE, {})
+                .get(direction.value, [])
+            )
+
+            for edge in edges:
+                if _DGQLFields.SOURCE in edge and _DGQLFields.TARGET in edge:
+                    rows.append(
+                        (
+                            edge[_DGQLFields.SOURCE],
+                            edge[_DGQLFields.TARGET],
+                            direction,
+                            current_depth,
+                        )
+                    )
+
+        return rows
+
+    def _recursive_trace(
+        self,
+        object_name: str,
+        object_domain: str,
+        direction: str,
+        current_depth: int,
+        total_depth: int,
+        object_version: Optional[str] = None,
+    ) -> List[Tuple[VariantType, VariantType, StringType, int]]:
+        """
+        Recursively traces lineage by making successive DGQL queries based on response nodes.
+        """
+        if current_depth > total_depth:
+            return []
+
+        lineage_edges = self._get_lineage(
+            object_name, object_domain, [direction], object_version, current_depth
+        )
+        if (
+            current_depth == total_depth
+            or not lineage_edges
+            or self._is_masked_object(lineage_edges[0][0])
+            or self._is_masked_object(lineage_edges[0][1])
```

Comment:
how about deleted objects? if a deleted object is returned, and we need request the next level of it. It seems you are using name to construct the request, but using name for a deleted obejcts might not be able to resolve. For example, for UI, we are using id field for requesting for recursive search

Reply:
Should we even go to the next level from the deleted node ?

Reply:
Discussed offline. For PrPr, we will not explore from deleted nodes. we can extend this based on user feedback.